filecoin-proofs/src/api.rs

```diff
@@ -481,3 +589,121 @@ fn pad_safe_fr(unpadded: &FrSafe) -> Fr32Ary {
     res[0..31].copy_from_slice(unpadded);
     res
 }
+
+#[test]
+fn test_generate_piece_commitment() -> Result<(), failure::Error> {
+    fn blah(data: &[u8]) -> Result<Commitment, failure::Error> {
+        let mut file = NamedTempFile::new().expects("could not create named temp file");
+        file.write_all(data)?;
+        let (comm_p, _) =
+            generate_piece_commitment(file.path(), UnpaddedBytesAmount(data.len() as u64))?;
+        Ok(comm_p)
+    }
+
+    {
+        // this scope tests comm_p generation for all byte lengths up to the minimum piece
+        // alignment when writing a piece to a sector
+        let max = 127;
+
+        for n in 0..=max {
+            let bytes: Vec<u8> = (0..n).map(|_| rand::random::<u8>()).collect();
+            let mut data_a = vec![0; n];
+            let mut data_b = vec![0; max];
+
+            for i in 0..n {
+                data_a[i] = bytes[i];
+                data_b[i] = bytes[i];
+            }
+
+            let comm_p_a = blah(&data_a)?;
+            let comm_p_b = blah(&data_b)?;
+
+            assert_eq!(comm_p_a, comm_p_b);
+        }
+    }
+
+    {
+        // this scope is just a sanity check that larger byte lengths are still zero padded
+        let bytes: Vec<u8> = (0..400).map(|_| rand::random::<u8>()).collect();
+        let mut data_a = vec![0; 400];
+        let mut data_b = vec![0; 508];
+
+        for i in 0..400 {
+            data_a[i] = bytes[i];
+            data_b[i] = bytes[i];
+        }
+
+        let comm_p_a = blah(&data_a)?;
+        let comm_p_b = blah(&data_b)?;
+
+        assert_eq!(comm_p_a, comm_p_b);
+    }
+
+    Ok(())
+}
+
+#[test]
+fn test_pip_lifecycle() -> Result<(), failure::Error> {
+    use crate::types::SectorSize;
+
+    fn add_piece<R, W>(
```

Comment:
TODO: Move this function declaration out of the `test_pip_lifecycle` function body.